src/tools/tidy/src/no_merge.rs

```diff
@@ -0,0 +1,77 @@
+//! This check makes sure that no accidental merge commits are introduced to the repository.
+//! It forbids all merge commits that are not caused by rollups/bors or subtree syncs.
+
+use std::process::Command;
+
+pub fn check(_: (), bad: &mut bool) {
+    let Ok(remote) = get_rust_lang_rust_remote() else { return; };
+    let Ok(merge_commits) = find_merge_commits(&remote) else { return; };
+
+    let mut bad_merge_commits = merge_commits.lines().filter(|commit| {
+        !(
+            // Bors is the ruler of merge commits.
+            commit.starts_with("Auto merge of")
+                || commit.starts_with("Rollup merge of")
+                // Our fellow subtrees are fine as well.
+                // This can be extended for future subtrees.
+                || commit.contains("Merge from rustc")
+                || commit.contains("merge rustc history")
+                || commit.contains("rust-analyzer")
+                || commit.contains("clippy")
+        )
+    });
+
+    if let Some(merge) = bad_merge_commits.next() {
+        tidy_error!(
+            bad,
+            "found a merge commit in the history: `{merge}`.
+To resolve the issue, see the merge commit help at the bottom of this post: https://jyn.dev/2022/09/02/git-cheats.html.
+If you're doing a subtree sync, add your tool to the list in the code that emitted this error."
+        );
+    }
+}
+
+/// Finds the remote for rust-lang/rust.
+/// For example for these remotes it will return `upstream`.
+/// ```text
+/// origin  https://github.com/Nilstrieb/rust.git (fetch)
+/// origin  https://github.com/Nilstrieb/rust.git (push)
+/// upstream        https://github.com/rust-lang/rust (fetch)
+/// upstream        https://github.com/rust-lang/rust (push)
+/// ```
+fn get_rust_lang_rust_remote() -> Result<String, String> {
+    let mut git = Command::new("git");
+    git.args(["remote", "--verbose"]);
+    let output = git.output().map_err(|err| format!("{err:?}"))?;
+    if !output.status.success() {
+        return Err("failed to execute git remote command".to_owned());
+    }
+
+    let stdout = String::from_utf8(output.stdout).map_err(|err| format!("{err:?}"))?;
+
+    let rust_lang_remote = stdout
+        .lines()
+        .find(|remote| remote.contains("rust-lang") && remote.contains("(push)"))
+        .ok_or_else(|| "rust-lang/rust remote not found".to_owned())?;
+
+    let remote_name = rust_lang_remote
+        .split_ascii_whitespace()
+        .next()
+        .ok_or_else(|| "remote name not found".to_owned())?;
+    Ok(remote_name.into())
+}
+
+/// Runs `git log --merges --format=%s $REMOTE/master..HEAD` and returns all commits
+fn find_merge_commits(remote: &str) -> Result<String, String> {
+    let mut git = Command::new("git");
+    git.args(["log", "--merges", "--format=%s", &format!("{remote}/master..HEAD")]);
```

Comment:
```
git rev-parse --abbrev-ref "master@{upstream}"
```
gives back `origin/master` for me, which is my outdated fork and not rust-lang/rust. Granted, this does work but it may be slower, especially if the fork master is very outdated.